client/src/includes/tabs.js

```diff
@@ -0,0 +1,118 @@
+    this.render = this.render.bind(this);
+
+    this.tabs = Array.from(this.el.querySelectorAll(this.options.tabLinkSelector));
+
+    this.activeTab = this.tabs[this.options.initialActiveTab];
+
+    this.tabs.forEach(tab => tab.addEventListener('click', this.setActiveTab));
+
+    // If there's a tab that already has the active class, activate it instead
+    const activeClassTab = this.el.querySelector(`${this.options.tabLinkSelector}.${this.options.tabActiveClass}`);
+    if (activeClassTab) {
+      this.activeTab = activeClassTab;
+    }
+
+    // If there's a tab that matches the location hash, activate it instead
+    if (window.location.hash) {
+      const activeHashTab = this.el.querySelector(`[href="${window.location.hash}"]`);
+      if (activeHashTab) {
+        this.activeTab = activeHashTab;
+      }
+    }
+
+    // Run render to ensure attributes are all properly set
+    this.render();
+  }
+
+  /**
+   * setActiveTab changes the currently active tab in an interface. It accepts
+   * three argument types:
+   * - an integer index
+   * - the tab DOM element
+   * - an Event triggered by the tab DOM element
+   */
+  setActiveTab(arg) {
+    let targetTab;
+
+    // Handle these cases: arg is event-like, arg is integer index, arg is element
+    if (arg instanceof Event) {
+      // Extract DOM element from event
+      targetTab = arg.target;
+    } else if (typeof arg === 'number') {
+      // Use arg as in index
+      targetTab = self.tabs[arg];
+    } else {
+      // Assume it's a DOM element
+      targetTab = arg;
+    }
+
+    // If the clicked tab is already active, bail out
+    if (this.activeTab === targetTab) return;
+
+    this.activeTab = targetTab;
+    window.history.replaceState(null, null, this.activeTab.href);
+    this.render();
+
+    // If arg is an event, stop propagation and prevent default behavior
+    if (arg instanceof Event) {
+      arg.stopPropagation();
+      arg.preventDefault();
+    }
+  }
+
+  render() {
+    this.tabs.forEach(tab => {
+      let pane;
+      if (tab === this.activeTab) {
+        // Add active class and ARIA attribute to active tab
+        tab.classList.add(this.options.tabActiveClass);
+        tab.setAttribute('aria-selected', 'true');
+        // Show active pane
+        pane = document.getElementById(tab.getAttribute('aria-controls'));
+        if (pane) {
+          pane.classList.add(this.options.paneActiveClass);
+          pane.hidden = false;
+        }
+      } else {
+        // Remove active class and ARIA attribute from any inactive tab
+        tab.classList.remove(this.options.tabActiveClass);
+        tab.setAttribute('aria-selected', 'false');
+        // Hide inactive panes
+        pane = document.getElementById(tab.getAttribute('aria-controls'));
+        if (pane) {
+          pane.classList.remove(this.options.paneActiveClass);
+          pane.hidden = true;
+        }
+      }
+    });
+  }
+}
+
+const initDefaultTabs = () => {
```

Comment:
What does the `default` in the function name stand for?

Reply:
I wanted to leave open the possibility that we might eventually want to call (or allow third-party devs to call) this on tabs that don't have the `tab-nav` class on it, so I'm considering the `tab-nav` tabs the default ones. No strong attachment to the naming here though.